[CUBRIDQA-937] add shell guide (#557)
CUBRID/cubrid-testtools: doc/shell_guide.md

Shell test cases contain almost all the feature test and some performance test which cannot be tested by sql test or other test suites.   
For examples:
* check whether a cubrid utility works well  
* check whether the error message of a sql statement is expected  
* check a specific cubrid parameter works

Shell test case path: [https://github.com/CUBRID/cubrid-testcases-private-ex/tree/develop/shell](https://github.com/CUBRID/cubrid-testcases-private-ex/tree/develop/shell)  
Shell test is contained by daily regression test. The test cases are run on both linux platform and Windows platform. This guide will introduce linux shell test. And shell test is also an important part of code coverage test. 

# 2 Shell Test via CTP
CTP is the only test tool which is used to execute shell test cases.   
Source URL: [https://github.com/CUBRID/cubrid-testtools](https://github.com/CUBRID/cubrid-testtools)  
## 2.1 Quick Start
**1. Install CUBRID**  
Please follow CUBRID installation guide to install CUBRID.  
Suppose that we install CUBRID at `$HOME/CUBRID`.  

> Then add `'source .cubrid.sh'` in `~/.bash_profile`:  

Reviewer: ```suggestion
Then append below into `~/.bash_profile`:  
```

shell/_06_issues/_18_1h/bug_bts_12583
```

### Execute Shell Test
Login controller node and run CTP:
```
cd ~/CTP/bin
nohup ./ctp.sh shell -c ~/CTP/conf/shell.conf &
```

### Check Test Results
Test results can be found in these ways:
1. the screen output (or file nohup.out)
2. feedback.log
3. qa homepage

#### feedback_type=file
If we set `feedback_type=file`, we can use file `'feedback.log'` to check the test results.  

#### feedback_type=database 
If we set `feedback_type=database`, we need check the test results on qahome page.  
For details, please refer to ['4.1 Verify regression test results'](#4.1-Verify-regression-test-results).  
This is the way we used in regression test.

#### Log File Introduction
|feedback_type=file|feedback_type=database|description|
|---|---|---|
|feedback.log|NA|Records the test result for each case, and the result summary. This file dose not exist if feedback_type=database.|
|test_status.data|NA|Records the current test status, such as total_case_count, total_fail_case_count. This file dose not exist if feedback_type=database.|
|main_snapshot.properties|main_snapshot.properties|Records the snapshot of all the parameters used in this test.|

> |check_local.log/check_106.log|check_local.log/check_106.log|Records the log of checking the environment.|

Reviewer: Change format as below in order fit table 

As is:
`check_106.log|check_local.log`

To be:
```
check_106.log  
check_local.log
```

|current_task_id|current_task_id|Records the main_id of this test. This is used when feedback_type=database. If feedback_type=file, '0' is written in this file.|
|dispatch_tc_ALL.txt|dispatch_tc_ALL.txt|Records all the cases that are needed to be test this time.|
|dispatch_tc_FIN_local.txt/dispatch_tc_FIN_106.txt|dispatch_tc_FIN_local.txt/dispatch_tc_FIN_106.txt|Records the cases that have been finished on this machine by now.|
|monitor_local.log/monitor_106.log|monitor_local.log/monitor_106.log|Records the monitor logs, such as memory, disk space, processes, CUBRID logs, database, CUBRID conf files, netstat. This kind of file is only used when we set enable_status_trace_yn=true in configuration file.|
|test_local.log/test_106.log|test_local.log/test_106.log|Records the screen output of CTP tool. It contains the sceen output of each test case.|

## 2.3 Execute A Single Test Case  

> Before execute a single case, we need export $init_path variable. Usually, we add it in `~/.bash_profile` file:    

Reviewer: ```suggestion
Before execute a single case, we need export `init_path` environmental variable. Usually, we add it in `~/.bash_profile` file:    
```

```
export init_path=$HOME/CTP/shell/init_path
```
To execute a single test case, we just need to login a test machine, go to the case path, and then execute shell command:  
```bash
cd /path/to/test case/
sh case_name.sh
 ```

# 3 Regression Test Deployment
## 3.1 Test Machines
|Role|User|IP|Hostname|
|---|---|---|---|
|controller node|shell_ctrl|192.168.1.104|func29|
|worker node|shell|192.168.1.104|func29|
|worker node|shell|192.168.1.105|func30|
|worker node|shell|192.168.1.106|func31|
|worker node|shell|192.168.1.107|func32|
|worker node|shell|192.168.1.108|func33|
|worker node|shell|192.168.1.109|func34|
|worker node|shell|192.168.1.110|func35|
|worker node|shell|192.168.1.111|func36|
|worker node|shell|192.168.1.112|func37|
|worker node|shell|192.168.1.113|func38|
|worker node|shell|192.168.1.114|func39|

## 3.2 Create and Set Users  
### Controller Node
We need create a new user: shell_ctrl.  
Then login root user and execute:  
```
sudo useradd shell_ctrl
```
Set password as our common password for user shell_ctrl.  
```
sudo passwd shell_ctrl
```
 Set the user's password to never expire.  
 ```
 sudo chage -E 2999-1-1 -m 0 -M 99999 shell_ctrl
 ```
 
### Worker Nodes

> We need create a new user: shell.  

Reviewer: ```suggestion
We need create a new user `shell`.  
```

Login root user and execute:  
```
sudo useradd shell
```
Set password as our common password for user shell.  
```
sudo passwd shell
```
 Set user's password to never expire.  
 ```
 sudo chage -E 2999-1-1 -m 0 -M 99999 shell
 ```
## 3.3 Install Software Packages

> Required software packages: jdk, lcov, bc.   

Reviewer: ```suggestion
Required software packages:
```

testcase_update_yn=true
testcase_git_branch=develop
testcase_timeout_in_secs=604800
test_platform=linux
test_category=shell
testcase_exclude_by_macro=LINUX_NOT_SUPPORTED
testcase_retry_num=0
delete_testcase_after_each_execution_yn=false
enable_check_disk_space_yn=true

owner_email=Mandy<cui.man@navercorp.com>

git_user=cubridqa
git_email=dl_cubridqa_bj_internal@navercorp.com
git_pwd=N6P0Sm5U7h

feedback_type=database

feedback_db_host=192.168.1.86
feedback_db_port=33080
feedback_db_name=qaresu
feedback_db_user=dba
feedback_db_pwd=

feedback_notice_qahome_url=http://192.168.1.86:8080/qaresult/shellImportAction.nhn?main_id=<MAINID>
```
shell_template.conf will be copied to \~/CTP/conf/shell_runtime.conf when test is started.  

3. Set `~/.bash_profile`  
Set `~/.bash_profile` like this:  
```
# .bash_profile

# Get the aliases and functions
if [ -f ~/.bashrc ]; then
        . ~/.bashrc
fi

# User specific environment and startup programs

PATH=$JAVA_HOME/bin:$HOME/CTP/common/script:$PATH:$HOME/.local/bin:$HOME/bin

export PATH

export CTP_BRANCH_NAME="develop"
export CTP_SKIP_UPDATE=0
```

> 4. Create a script to start consumer

Reviewer: ```suggestion
4. Create a script to start consumer  
```

1. Install CTP
Please refer to ["CTP Installation Guide"](https://github.com/CUBRID/cubrid-testtools/blob/develop/doc/ctp_install_guide.md#3-install-ctp-as-regression-test-platform)

2. Set `~/.bash_profile`  
Set `~/.bash_profile` like this:  
```
# .bash_profile

# Get the aliases and functions
if [ -f ~/.bashrc ]; then
        . ~/.bashrc
fi

# User specific environment and startup programs

PATH=$PATH:$HOME/.local/bin:$HOME/bin

export CTP_HOME=$HOME/CTP
## init_path is used when we run shell case manually on this machine.
export init_path=$CTP_HOME/shell/init_path

export PATH=$CTP_HOME/bin:$CTP_HOME/common/script:$PATH

export CTP_BRANCH_NAME="develop"
export CTP_SKIP_UPDATE=0

. ~/.cubrid.sh
export GCOV_PREFIX=/home/shell
export GCOV_PREFIX_STRIP=2
ulimit -c unlimited
```

3. Deploy test cases
```
git clone --no-checkout https://github.com/CUBRID/cubrid-testcases-private-ex.git
cd ~/cubrid-testcases-private-ex
git config core.sparseCheckout true
echo 'shell/*' > ~/cubrid-testcases-private-ex/.git/info/sparse-checkout
git checkout develop
```

4. Make directories for test
```
cd
mkdir do_not_delete_core
mkdir ERROR_BACKUP
```

5. Add `.cubrid.sh` file   

> If cubrid has never been installed on the machine, we need add file '.cubrid.sh' at $HOME path manually.  

Reviewer: ```suggestion
If cubrid has never been installed on the machine, we need add file `'.cubrid.sh'` at `$HOME` path manually.  
```

## 3.6 Start Consumer

> On controller node, execute the script start_test.sh to start listening the test message after deployment.  

Reviewer: ```suggestion
On controller node, execute the script `~/start_test.sh` to start listening the test message after deployment.  
```

> This will start a shell test when the consumer receive the test message.

Reviewer: ```suggestion
This will start a shell test when the consumer receives the test message.
```

```
cd ~
sh start_test.sh
```

# 4 Regression Test Sustaining
We execute shell test for each CI build, and execute code coverage test monthly.  

> Both of the test are started automatically when the controller receive a test message. We just need to verify the test results, and report the issues.

Reviewer: ```suggestion
Both of the tests are started automatically when the controller receives a test message. We just need to verify the test results, and report the issues.
```

## 4.1 Verify Regression Test Results
### Result Overview

> Open qahome in browser, then select build number. 'Function Result' page will be displayed.

Reviewer: ```suggestion
Open [QA homepage](http://qahome.cubrid.org) in web browser, then select build number. `'Function'` tab page will be displayed.
```

Find `shell` and `shell_debug` in linux part, and find `shell` in windows part.  
![shell_linux](./shell_image/image1.png)  
![shell_windows](./shell_image/image2.png)  
If there are crash failures, an alarm icon will appear in the 'Total' column, like this:  
![shell_crash](./shell_image/image12.png)     

> In the overview page, we should check the values in all the colomns. The most important check items are:   

Reviewer: Check below format.

Reviewer: Again to check.

|column|check point|
|---|---|
|Test Rate|Test rate should be 100%.|
|Fail|How many cases are failed.|
|Verified Rate|After the verification, verified rate should be 100%.|
|Elapse Time|The elapse time of the test should not be longer than the elapse time of the previous build too much.|   

### Verfy the Failed Cases
Click the link in this image to open the online verification page.  
![online_verify1](./shell_image/image3.png)   
On the verification page, click 'Log in' to login.  
![online_verify2](./shell_image/image4.png)  
After logging in, we can start to verify the failed cases online.    
![online_verify3](./shell_image/image5.png)    
**Click the link in 'Case File' column to check the case and the screen output.**  
![online_verify6](./shell_image/image8.png)  
`'[history]':` check the test case history failure   
*history page*    
![online_verify4](./shell_image/image6.png)   
`'[verify]':` open the case's verification page to write the verification result   
*case verification page*   
![online_verify5](./shell_image/image7.png)  
On this page:
1. select a failure reason.  

> 2. write some comments in 'Reason content'  

Reviewer: Check all like below:

```suggestion
2. write some comments in `'Reason content'` 
```

3. write the issue number in 'New issues' or 'New issues'  
4. click the button at the bottom to close this page.  
Fix: I am sure of this verification result.  
Save: I am not sure of this verification result. I will decide it later.  

Sometimes, we cannot find the failure reason from the screen output directly. At this time, we need login the test machine, go to the case path, check the logs. If we still cannot find the failure reason from the logs, we should run the case step by step on this machine to find the failure reason.  
We should verify the crash failures first.  

### Verify the Low Performance Test Cases

> Open qahome in browser, then select build number. `'Function Result'` page will be displayed. Find `'Low Performance Scenarios'` link.  

Reviewer: ```suggestion
Open [QA homepage](http://qahome.cubrid.org) in web browser, then select build number.  `'Function'` tab page will be displayed. Find `'Low Performance Scenarios'` link.  
```

![online_verify6](./shell_image/image13.png)   
Click this link, and go to the `'Low Performance Scenarios'` page.  
Click 'SHELL', to check the shell test cases.   
![online_verify7](./shell_image/image14.png)   
We can change the filters, such as 'OS', 'Bits', 'Category', 'Base build'.  
Usually, we use the last CI build as base build to compare with.    
The elapse time of these cases should not changed too much compared to the base build.  

> And a red icon '+' will be displayed in 'Flag' column if the reuslt is abnormal. 

Reviewer: ```suggestion
And a red icon '+' will be displayed in 'Flag' column if the result is abnormal. 
```

## 4.2 Verify Code Coverage Test Result
Go to QA homepage and find the 'code coverage' node in the left area, click the link of the latest result.  
![code coverage 1](./shell_image/image19.png)  
Click `'shell'` link.  
![code coverage 2](./shell_image/image10.png)   

> Check the value of 'Coverage' column. The coverage value should not lower than previous code coverage test result.  

Reviewer: Insert empty line for easy read.
Check all such cases.

#these parameters are used when feedback_type=database
feedback_db_host=192.168.1.86
feedback_db_port=33080
feedback_db_name=qaresu
feedback_db_user=dba
feedback_db_pwd=
```
### Excluded List  
The cases in the excluded list will not be run in the test.  
If the case will block the test (eg. it hangs in regression test and the issue will not be fixed recently), we should add the case to the excluded list. For shell test, we have two excluded list files:
```
shell/config/daily_regression_test_excluded_list_linux.conf
shell/config/daily_regression_test_excluded_list_windows.conf
```
For example, if we need add case `'shell/_06_issues/_18_1h/bug_bts_12583'` in linux excluded list, we should add these lines in file `'shell/config/daily_regression_test_excluded_list_linux.conf'`.  
```
#CBRD-21358 (add comment in this line to record the reason of adding this case in the exclude list)
shell/_06_issues/_18_1h/bug_bts_12583
```

### Execute Shell Test
Login controller node and run CTP:
```
cd ~/CTP/bin
nohup ./ctp.sh shell -c ~/CTP/conf/shell.conf &
```

### Check Test Results
Test results can be found in these ways:
1. the screen output (or file nohup.out)
2. feedback.log

> 3. qa homepage

Reviewer: ```suggestion
3. [QA homepage](http://qahome.cubrid.org)
```

#### feedback_type=file
If we set `feedback_type=file`, we can use file `'feedback.log'` to check the test results.  

#### feedback_type=database 
If we set `feedback_type=database`, we need check the test results on qahome page.  

> For details, please refer to ['4.1 Verify regression test results'](#4.1-Verify-regression-test-results).  

Reviewer: Check the link, please.

This is the way we used in regression test.

#### Log File Introduction
|feedback_type=file|feedback_type=database|description|
|---|---|---|
|feedback.log|NA|Records the test result for each case, and the result summary. This file dose not exist if feedback_type=database.|
|test_status.data|NA|Records the current test status, such as total_case_count, total_fail_case_count. This file dose not exist if feedback_type=database.|
|main_snapshot.properties|main_snapshot.properties|Records the snapshot of all the parameters used in this test.|
|check_local.log<br>or<br>check_106.log|check_local.log<br>or<br>check_106.log|Records the log of checking the environment.|
|current_task_id|current_task_id|Records the main_id of this test. This is used when feedback_type=database. If feedback_type=file, '0' is written in this file.|

> |dispatch_tc_ALL.txt|dispatch_tc_ALL.txt|Records all the cases that are needed to be test this time.|

Reviewer: ```suggestion
|dispatch_tc_ALL.txt|dispatch_tc_ALL.txt|Records all the cases that are needed to be tested this time.|
```

3. write the issue number in `'New issues'` or `'Revise required issues'`  
4. click the button at the bottom to close this page.  
`Fix`: I am sure of this verification result.  
`Save`: I am not sure of this verification result. I will decide it later.  

Sometimes, we cannot find the failure reason from the screen output directly. At this time, we need login the test machine, go to the case path, check the logs. If we still cannot find the failure reason from the logs, we should run the case step by step on this machine to find the failure reason.  
We should verify the crash failures first.  

### Verify the Low Performance Test Cases
Open ['QA homepage'](http://qahome.cubrid.org) in browser, then select the build number. `'Function Result'` page will be displayed.   
Find `'Low Performance Scenarios'` link.  

![online_verify6](./shell_image/image13.png)   

Click this link, and go to the `'Low Performance Scenarios'` page.  
Click 'SHELL', to check the shell test cases.   

![online_verify7](./shell_image/image14.png)   

We can change the filters, such as `'OS'`, `'Bits'`, `'Category'`, `'Base build'`.  
Usually, we use the last CI build as base build to compare with.    

> The elapse time of these cases should not changed too much compared to the base build.  

Reviewer: ```suggestion
The elapse time of these cases should not change too much compared to the base build.  
```

> And a red icon `'+'` will be displayed in `'Flag'` column if the result is abnormal. 

Reviewer: The read icon means the elaspe time in compared build is under threshold, but on current build, it as new one is great than threshold.

![code coverage 3](./shell_image/image11.png) 

## 4.3 Report Issues  
### Crash Issue  
If there is a crash, report it following the rules in this link: [‘How to Report Regression Crash Issues’](http://jira.cubrid.org/browse/CUBRIDQA-1?focusedCommentId=4739244&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-4739244)  

### Normal Issue  
If the the test result is as expected, first we should confirm whether it is revise required. If not, report it as an issue on jira. 

## 4.4 Execute A Shell Test by Sending Message
We can use the regression tools to trigger a test.  
1. Confirm (and change the parameters if needed) in `'~/CTP/conf/shell_template.conf'` on controller node.    
2. Send a message to start the test  
Login: message@192.168.1.91  
Send test message as:  
```
sender.sh QUEUE_CUBRID_QA_SHELL_LINUX http://192.168.1.91:8080/REPO_ROOT/store_01/10.2.0.8330-d4d8464/drop/CUBRID-10.2.0.8330-d4d8464-Linux.x86_64.sh shell default
```
3. Start consumer on the controller node  
If the consumer is not started, login the controller node to start it.  
```
cd ~
sh start_test.sh 
```
4. Check test result  

> The results will be uploaded to qahome automatically. You can follow ['4.1 Verify regression test results'](#4.1-Verify-regression-test-results) to check the test results.   

Reviewer: Confirm the link worked.

Please check all links.

## 4.5 Execute A Code Coverage Test
For code coverage test, just need to send a message.  
Login: message@192.168.1.91  
Send test message as:  
```
cd ~/manual

> sh sender_code_coverage_testing_message.sh Queue:QUEUE_CUBRID_QA_SHELL_LINUX Build URL:http://192.168.1.91:8080/REPO_ROOT/store_01/10.2.0.8270-c897055/drop/CUBRID-10.2.0.8270-c897055-gcov-Linux.x86_64.tar.gz Source URL:http://192.168.1.91:8080/REPO_ROOT/store_01/10.2.0.8270-c897055/drop/cubrid-10.2.0.8270-c897055-gcov-src-Linux.x86_64.tar.gz Category:shell

Reviewer: Check the command worked.

cbrd_xxxxx_1/cases/cbrd_xxxxx_1.sh
cbrd_xxxxx_{issue_key}/cases/cbrd_xxxxx_{issue_keyword}.sh
```
Example:
```
cubrid-testcases-private-ex/shell/_06_issues/_19_2h/cbrd_22586/cases/cbrd_22586.sh
```

## 5.2 Shell Case Template
```
#!/bin/sh
# Initialize the environment variables, and import all the functions. 
. $init_path/init.sh
init test

dbname=tmpdb

cubrid_createdb $dbname

# Test steps

# Check the result
if [condition]
then
       write_ok
else
       write_nok
fi

cubrid service stop

# The command 'cubrid deletedb' will check whether there are core files and fatal error generated in the case, and backup db volumns, core files, logs.  
cubrid deletedb $dbname

# Clean environment, such as delete the jave class files

# 'finish' is a function in init.sh, which will revert all the conf files to the original status.
finish

``` 

## 5.3 `'cubrid'` Script
When execute `'init test'` at the beginning of the case, `'${init_path}/../../bin:${init_path}/../../common/script'` is added to PATH:  
```
PATH=${init_path}/../../bin:${init_path}/../../common/script:$PATH
```
`'cubrid'` script in init_path will be used instead of $CUBRID/bin/cubrid in the case. For example:  

> `'cubrid deletedb':`  

Reviewer: ```suggestion
*'cubrid deletedb':*  
```

It will check whether there are core files and fatal error generated in the case. If there are, backup db volumns, core files, logs. Then execute $CUBRID/bin/cubrid deletedb.  

> `'cubrid checkdb':`  

Reviewer: ```suggestion
*'cubrid checkdb':*  
```
